Fixes #25335 - Creating a React/Redux Editor

glekner opened the pull request:
[Discourse](https://community.theforeman.org/t/creating-a-new-redux-react-editor/11627) to discuss new UX

_TODO_

- [x] Preview Action

- [x] Tests

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

amirfefer commented on webpack/assets/javascripts/react_app/components/Editor/Editor.js:
```diff
@@ -0,0 +1,119 @@
+    }
+    if (isMasked && type === 'templates') changeState('isMasked', false);
+    if (selectedView !== 'input') changeState('selectedView', 'input');
+    if (isRendering) changeState('isRendering', false);
```
wouldn't it be better to use some sort of initialization action?

glekner replied:
Great suggestion

amirfefer commented on webpack/assets/javascripts/react_app/components/Editor/EditorReducer.js:
```diff
@@ -0,0 +1,51 @@
+  switch (action.type) {
+    case EDITOR_CHANGE_STATE: {
+      const [key] = Object.keys(payload);
+      if (Object.prototype.hasOwnProperty.call(state, key)) {
```
how about moving this logic into the action (why firing an action if it doesn't change the state?) , plus throw an error if it doesn't exist.

glekner replied:
Great suggestion 2#

glekner commented:
@theforeman/ui-ux 
I removed WIP label, this PR is getting mature and the review process should start.

Feedback I need:
- Test the Preview function (is it working properly?)
- What code can be deprecated/removed ? 

Editor views you can try:
- Monitor => Report Templates
- Host => Partition, Provisioning Templates
- Configure => Global Parameters

ohadlevy commented on app/views/common_parameters/_form.html.erb:
```diff
@@ -5,21 +5,20 @@
-
-        </div>
+            <div id="react-ace"></div>
+              <%= mount_react_component('Editor', "#react-ace", {
```
would this work if you have more than one parameter? I believe each div needs to have a unique id.

glekner replied:
this form is being rendered when you click on one parameter, I have multiple parameters and it works fine.

ohadlevy replied:
thanks, I misunderstood this to be the domain parameters (or other parameters) where we dynamically add forms to the UI, can you validate those are not using the same code path?

ohadlevy commented on app/helpers/editor_helper.rb:
```diff
@@ -0,0 +1,13 @@
+
+  def template_name_attribute(template)
+    return 'provisioning_template[template]' if template == 'ProvisioningTemplate'
+    return 'report_template[template]' if template == 'ReportTemplate'
```
I believe this can be extracted from the object itself,  perhaps using https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-fields_for ?

ohadlevy commented on test/integration/report_template_js_test.rb:
```diff
@@ -7,33 +7,6 @@ class ReportTemplateJSIntegrationTest < IntegrationTestWithJavascript
       assert_index_page(report_templates_path, "Report Templates", "Create Template")
     end
   end
-  test "creating report templates with inputs, displaying them when generating the template" do
```
why delete all of the tests? I would assume they are still valid?

glekner replied:
The deleted tests are trying to change the editor value, but its not rendered in the html.
I tried putting `wait_for_ajax` but still the value of the `@response` is:

`//<![CDATA[
$(tfm.reactMounter.mount('Editor', '#react-ace', { ...`

If theres a way to fix it then those tests can be fixed.
Anyway, i'm testing the editor in the JS tests.

ohadlevy commented on package.json:
```diff
@@ -88,6 +88,7 @@
     "patternfly-react": "^2.19.1",
     "prop-types": "^15.6.0",
     "react": "^16.4.0",
+    "react-ace": "^6.2.0",
```
[EDIT: PBKAC] packaging PR at https://github.com/theforeman/foreman-packaging/pull/3207

glekner commented on test/integration/common_parameter_test.rb:
```diff
@@ -13,15 +15,10 @@ class CommonParameterIntegrationTest < ActionDispatch::IntegrationTest
-    assert page.has_content? 'mynewvalue'
-  end
-
-  test "does not display editor on hidden value" do
```
this functionality is being tested in the JS tests

glekner commented:
Fixed rb integration tests

glekner commented:
Deleted all JQuery from `foreman_editor.js` 🎉

ohadlevy commented on test/integration/parameters_permissions_domain_test.rb:
```diff
@@ -16,48 +16,6 @@ class ParametersPermissionsIntegrationTest < ActionDispatch::IntegrationTest
-        click_link 'Delete'
-      end
-
-      assert page.has_no_css? "tr#common_parameter_#{@visible_global_parameter.id}_row"
```
these are good tests to keep...?

ohadlevy commented on test/integration/parameters_permissions_global_js_test.rb:
```diff
@@ -0,0 +1,60 @@
+      click_link @visible_global_parameter.name
+      find('#editor').click
+      find('.ace_content').send_keys "another_value"
+      sleep 1 # Wait for the editor onChange debounce
```
i assume wait_for_ajax doesnt work? /cc @amirfefer

ohadlevy replied:
@amirfefer can you have a look please?

ohadlevy commented on webpack/assets/javascripts/react_app/components/Editor/Editor.fixtures.js:
```diff
@@ -0,0 +1,56 @@
+};
+
+export const dropdowns = {
+  modes: ['Text', 'Json', 'Ruby', 'Html_ruby', 'Sh', 'Xml', 'Yaml'],
```
why is this defined twice? just wondering why we need it in the rails part to begin with?

glekner replied:
these are just the fixtures for testing. But the real data do come from the rails, do we prefer it in the client?

ohadlevy replied:
if its static list, then yes?

ohadlevy replied:
import the fixtures from the const file? (instead of defining twice?)

glekner commented:
@ares Can you try the Editor? would like feedback :)

ohadlevy commented:
@glekner can you rebase please?
@ares I would really love to get this merged soon, will you have time to test out the implementation ?

glekner commented:
@ohadlevy I rebased yesterday, this is ready for testing. Travis w/node 6 fails here like in other PRs I saw.

sharvit commented on webpack/assets/javascripts/react_app/components/Editor/Editor.js:
```diff
@@ -0,0 +1,170 @@
+class Editor extends React.Component {
+  componentDidMount() {
+    const { initializeEditor } = this.props;
+    initializeEditor(this.props);
```
Could you send only the relevant data? It feels risky and we might lose control over what actually sent to this method.

glekner replied:
can i access the state from the action itself with a selector? is it a good idea?

sharvit replied:
You can access the state from within the action and pass it to a selector (which is one of the reason I try to push us back to selectors that recive the whole state object).

Some developers says you never actually need to do that because you can recive the state into your component props (with selectors) so you can just pass them to your actions.
I bealive in some cases it is more beneficial to do what you described and use the `getState` method.

In this case, do you think it would be more beneficial to get your data directly from the state? Isn't all the data you need already exists for you in the component props?

sharvit commented on webpack/assets/javascripts/react_app/components/Editor/EditorActions.js:
```diff
@@ -0,0 +1,127 @@
+    initialState.hosts = Object.assign(
+      {},
+      ...hosts.map(element => ({ [element.host.name]: element.host.id }))
+    );
```
What's the difference between:

```js
initialState.hosts = {
  ...hosts.map(element => ({ [element.host.name]: element.host.id }))
};
```

glekner replied:
This won't work, as it returns an object of objects e.g.
```
 {
   0: {aaron-ellert.ramon-eatherly.gail-walema.example.org: 22}
   1: {abel-bearman.ruben-montieth.example.net: 66}
   2: {alvin-bocchini.erika-kirton.elton-lucore.example.org: 44}
 }
```

my code returns:

```
 {
   aaron-ellert.ramon-eatherly.gail-walema.example.org: 22
   abel-bearman.ruben-montieth.example.net: 66
   alvin-bocchini.erika-kirton.elton-lucore.example.org: 44
 }
```

sharvit commented on webpack/assets/javascripts/react_app/components/Editor/EditorActions.js:
```diff
@@ -0,0 +1,127 @@
+
+  dispatch({
+    type: EDITOR_INITIALIZE,
+    payload: { initialState },
```
Can simply be: `payload: initialState`

sharvit replied:
It's a pattern in all your actions, it's kind of nitpicking, but I see it as redundant.

sharvit commented on webpack/assets/javascripts/react_app/components/Editor/EditorActions.js:
```diff
@@ -0,0 +1,127 @@
+  });
+};
+
+export const changeState = newState => (dispatch, getState) => {
```
I am pretty much against the idea of using generic actions and reducers as it goes against a lot of the principles of Redux.

See:
https://stackoverflow.com/questions/48308444/i-created-a-generic-reducer-for-the-state-when-should-it-be-used-and-when-shou
https://softwareengineering.stackexchange.com/questions/322222/creating-generic-actions-in-redux

sharvit commented on webpack/assets/javascripts/react_app/components/Editor/EditorActions.js:
```diff
@@ -0,0 +1,127 @@
+    );
+  }
+
+  dispatch({
```
You don't actually need the `dispatch`, can just return this object like in simple actions (it's not async action).

ohadlevy replied:
@glekner bump on this ?

sharvit commented on webpack/assets/javascripts/react_app/components/Editor/EditorActions.js:
```diff
@@ -0,0 +1,127 @@
+};
+
+export const revertChanges = template => dispatch => {
+  dispatch({
```
`dispatch` is redundant.

sharvit commented on webpack/assets/javascripts/react_app/components/Editor/EditorReducer.js:
```diff
@@ -0,0 +1,61 @@
+  switch (action.type) {
+    case EDITOR_CHANGE_STATE: {
+      return state.merge(payload.newState);
+    }
```
Can remove the `{` and the `}` and remove 6 lines

sharvit commented on webpack/assets/javascripts/react_app/components/Editor/components/EditorOptions.js:
```diff
@@ -0,0 +1,266 @@
+import { bindMethods } from '../../../common/helpers';
+import DiffRadioButtons from '../../DiffView/DiffRadioButtons';
+
+const cogPopover = data => (
```
Can't it be a component that lives in its own file?

sharvit reviewed (changes requested):
Awesome work @glekner :1st_place_medal:

glekner commented:
- Removed the generic action, and implemented most of @sharvit suggestions.

ohadlevy commented:
@glekner can you rebase please? recent changes to parameters imho have changed the editor implementation for global parameters.

@ares it would be great if you can review this asap, so further painful rebases can be avoided...

glekner commented:
Rebased. Looking at the `_form` changes of the parameters, looks like there is no need for an editor in global parameters anymore, so I removed it.

ohadlevy commented:
I wonder if it was removed because of the lack of ability to syntax
highlight or some other reason? @ares?

On Mon, Feb 25, 2019 at 12:30 PM Gilad Lekner <notifications@github.com>
wrote:

> Rebased. Looking at the _form changes of the parameters, looks like there
> is no need for an editor in global parameters anymore, so I removed it.
>
> @ohadlevy <https://github.com/ohadlevy>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/6180#issuecomment-466959449>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx4cWWtq1braJg8gT0kYvhM4NtKR9ks5vQ7ssgaJpZM4X_U0p>
> .
>

ares commented:
IIRC it didn't support the way we hide parameters that are set for hidden values. We're using font that changes all characters with dots. Therefore if global parameter contains a password and is marked as "hidden value", by opening the edit form, it does not show it to people looking over a shoulder.

Global parameters can contain erb and now since they can have complex type, the full editor can be useful again. Does it work with hidden values?

glekner commented:
@ares the new editor supports hiding a value, I added this feature specifically for this view. But it doesn't hide it on default, so for example if a user refreshes the page it will be visible, if we want to save the hidden state I guess we should save it in the db. so do we want to use the new editor there?

ares commented:
> if we want to save the hidden state I guess we should save it in the db. so do we want to use the new editor there?

I don't fully understand, the attribute is stored in DB so the editor should look at `$parameter.hidden_value` and hide or show the value based on this. I think having advanced editor for parameters would be even more important now when we have advanced types such as array or hash for both parameters and ansible variables. Note that params also can contain erb syntax.

But at the same time, most parameter values are one-line values. So the full editor should probably only open in fullscreen. Anyway, that can be separate PR.

I'm starting testing and will focus on templates editing only now.

ares reviewed (changes requested):
First of all, this looks as a great addition. I found few things from user point of view:

1) clicking preview, it does not render the preview until I select some host, it used to render it for the first host in the selector, which is a bit nicer as usually I'm not interested in specific host

2) if I go back to editing, change something and click preview, I see old preview, it's not clear I need to choose the preview host again to see fresh preview

3) the previous selector for host was nicer if I need to switch from one host to another, now I need to click the cross to clean the field and then click again to input to open the list

4) is the list paginated somehow? it looks it just takes preview_host_collection for a given class, that can be 10 000 hosts or more, this is why we had ajax to load matching hosts

5) ~changing SYNTAX in settings does not do anything, it remains as Text~, now I realized it changes to Text when I'm in preview, I can't select anything else then but when I switch back to edit, it changes to the value I selected in preview, which is slightly confusing, though probably acceptable if there's no good solution for this

6) labels in Settings are all capitals, should it be that way?

7) after opening full screen, there seems to be a white bar in the modal at bottom, is this intended?

8) rendering a preview can take some time, there's no indication something is loading

9) there's a white gap between the editor menu and editor text area, it used to be there previously too but not the menu feels more like tabs, perhaps it would look better without the white space?

10) import silently sets the content of editor while I'm in preview mode, maybe we should disable it in preview and change?

11) import file seems to only work once, second time it fails silently

One question - is it possible to specify format when I mount the component? E.g. for ansible playbooks, it would be great to set YAML right away. Not needed in this PR though :-) 

In particular, I like the fact it works with REX plugin out of the box. I like the new diff and also the new design of editor header. Good job. If you're interested I have couple of ideas how to improve the template editing further, but that's outside of scope for this PR :-)

ares commented on app/helpers/editor_helper.rb:
```diff
@@ -0,0 +1,7 @@
+  def template_name_attribute(template)
+    return 'provisioning_template[template]' if template == 'ProvisioningTemplate'
+    return 'report_template[template]' if template == 'ReportTemplate'
+    return 'ptable[template]' if template == 'Ptable'
```
how do we handle templates introduced by plugins (e.g. JobTemplate form REX plugin), I suppose this is used for form element name, could perhaps this work?
```
"#{template.class.underscore}[template]"
```

though I think "template" attribute will not always the name of attribute for which we mount the editor

glekner commented:
_Updates_ 🚀 
* Replaced TypeAhead with Select component for the host picker for preview.
* Preview is triggered automatically with the first available host upon going to the preview tab
* Settings labels are not uppercased anymore
* Added a spinner when a preview is loading
* Removed gap between the editor header and the content
* Disabled the Import button when not in the Editor view.

@ares regarding your review:

7. It depends on the longest line in the file, if its wider than the window you will see a scrollbar which is probably white.
10. Import works twice for me, can you explain how to reproduce this bug?

_TODO:_
1) Fetch hosts w/API
2) Warn the user the preview is outdated if he changed the template and navigated to the Preview tab again.

glekner commented:
_Updates_ 🚀 
* Added a Warning whenever a rendered preview thats displayed is outdated.
* Disabled the options to change Syntax/Keybind on the Preview tab.
* Clearing the host selection will also clear the preview

@ohadlevy @ares  Do you have maybe an advice on how to fetch hosts w/API
that are returned from [this function](https://github.com/theforeman/foreman/blob/develop/app/controllers/hosts_controller.rb#L623)
I guess what is needed is similar to what [we did](https://github.com/theforeman/foreman/pull/6469/files#diff-608d83c9bbc71d670e46c1a9cd1812eaR7) in the WIP audits PR, But ,maybe adding a `format.json` with the preview function to the host controller is too specific because its for templates only?

glekner commented:
- Added selectors
- Regarding the Select component, it supports an `ajax` func but it can't be attached to a `<select>` which is what the component use. this issue is fixed with Select v4 so I guess i can create another PR which upgrades it or I can even write a new Select component that doesn't use JQuery, anyway its a shame it blocks this PR but I guess its necessary.

ares commented:
this needs rebase now

Commit:
Fixes #25335 - Create a React/Redux Editor

glekner commented:
@ares this was not the branch I meant, but since i finished my work on it I merged it with this PR.
It's now updated, containing the files for the new Select component, which will be deleted next week, since the review is complete on the [PF side.](https://github.com/patternfly/patternfly-react/pull/1650)

This is now updated for you to test.

ares reviewed (changes requested):
Thanks for update @glekner 

I gave it another round, here are my findings:

1) pressing enter in editor does not add a new line

2) the first line seems to be missing some top padding
![editor_1](https://user-images.githubusercontent.com/109773/55624986-c81cdb00-57a7-11e9-8259-fa44e67de1da.png)

3) the new selector now works nicely as expected, can it be styled closer to select2? or could we modify select2 styles so that they look as similar as possible?

Functionality wise, it works really well now! I suppose we'll need to wait until pf-react PR gets in and is released and then we'll need to bump the version?

glekner commented:
thanks @ares 

1. fixed
2. You said we should remove the white space between the navbar and the editor. we can either bring it back but smaller, or just stay as it is?
3. awsome! about styling i guess we can change it later, and migrate between the two components. I styled it based on the PF design

ares commented:
For 2. it's good that there's no space between editor and navbar. I meanT padding inside the editor window. It feels like the line height of the first line is too small. I think in css language it's padding-top of the black editor text area that needs to be bigger.

glekner commented:
- removed `react-click-outside` package and implemented it myself in `EditorHostSelect`
- added padding-top for the first line of the editor

Now waiting for `patternfly-react-extensions` to release a new version, the Select component is already merged. @ares

glekner commented:
- rebased + using Select component from patternfly

I believe we are ready 🚀

glekner commented:
[test upgrade]
[test katello]

sharvit reviewed (dismissed):
Thanks for the long effort @glekner LGTM :1st_place_medal:

mmoll commented:
are there RPM PRs needed for this?

ohadlevy commented:
> are there RPM PRs needed for this?

I did try a long time ago at https://github.com/theforeman/foreman-packaging, probably needs to be revisited...

ares commented:
There are few more things that don't work
1) go to existing template, change the content then switch to preview, click on revert button and confirm, the editor activates `Editor tab` again but it seems preview is still there, I need to click Preview tab and back to Editor to get the new text content

2) when I click on preview after I updated the text in Editor, I see that it's out of date, that's nice but previously it recreated preview right-away this forces me to do 2 clicks, what's the reason for it? I like the styling of message though :-)

Otherwise works good. We'll need @theforeman/packaging ack though for new dependencies I guess. @glekner mind to try packaging new dependencies?

glekner commented:
nice finds @ares 

1. fixed
2. I guess I thought about scenarios where you want to go back and forth between the first preview and your code before updating the preview? maybe i'm wrong.

about packaging, i'll try :)

glekner commented on package.json:
```diff
@@ -91,9 +91,11 @@
+    "patternfly-react-extensions": "^2.18.2",
     "prop-types": "^15.6.0",
     "react": "^16.8.1",
+    "react-ace": "^6.3.2",
```
https://github.com/theforeman/foreman-packaging/pull/3706

ohadlevy replied:
i still think you need to remove the brace from this file, it is no longer imported by foreman (rather via by react-ace)

glekner replied:
@ohadlevy  we need it in `EditorView` - https://github.com/theforeman/foreman/pull/6180/files#diff-43445d906144300b418f2f9a8c540820R5

as stated in Ace docs - https://github.com/securingsincity/react-ace/blob/master/docs/Ace.md

ohadlevy replied:
does it make sense to export brace from react-ace? i think its wrong to require it twice (here and at brace) perhaps send them a pr to export it? then we wont have two copies of it...?

ohadlevy replied:
@glekner https://github.com/securingsincity/react-ace/issues/638 ?

ohadlevy replied:
ok, can you align the brace version again (so they dont conflict with react-ace?) i think this should be enough to unblock this PR. I already created a packaging PR at https://github.com/theforeman/foreman-packaging/pull/3706

glekner replied:
aligned

glekner commented on package.json:
```diff
@@ -91,9 +91,11 @@
     "patternfly": "^3.58.0",
-    "patternfly-react": "^2.30.6",
+    "patternfly-react": "^2.33.0",
+    "patternfly-react-extensions": "^2.18.2",
```
https://github.com/theforeman/foreman-packaging/pull/3707

ohadlevy commented on app/assets/stylesheets/base.scss:
```diff
@@ -182,10 +182,6 @@ select {
-.ace_editor {
-  border: 2px solid #eee;
-}
-
```
there are some leftover css classes such as .fullscreen* and similar function names in JS.

ohadlevy replied:
@glekner do you want to follow up with another PR after this one gets merge?

ohadlevy commented on package.json:
```diff
@@ -71,7 +71,7 @@
     "babel-polyfill": "^6.26.0",
     "bootstrap-sass": "^3.3.7",
-    "brace": "^0.10.0",
+    "brace": "^0.11.1",
```
is it required? I think its actually now bundled in react-ace and we can remove it from here?

ohadlevy commented on package.json:
```diff
@@ -91,9 +91,11 @@
     "number_helpers": "^0.1.1",
     "patternfly": "^3.58.0",
-    "patternfly-react": "^2.30.6",
+    "patternfly-react": "^2.33.0",
```
I think you need to package this one too? at the time of writing its 2.32.4  ? (not seeing 2.33?)

ohadlevy commented:
@ares there are minor code nitpicks that needs to be handled, but can you ack from a usability POV ? thanks!

ares commented:
I'll retest once code is final, there were regressions in last patches, but I don't expect any other findings. I see https://github.com/theforeman/foreman-packaging/pull/3707 has been filed :+1: 

@glekner please ping me when you push the new version, thanks!

glekner commented:
- removed unused css, reset brace version.

Packaging:
- PF-React - https://github.com/theforeman/foreman-packaging/pull/3752
- PF-Extensions - https://github.com/theforeman/foreman-packaging/pull/3707
- ReactAce - https://github.com/theforeman/foreman-packaging/pull/3706

ping @ares @ohadlevy

glekner commented on package.json:
```diff
@@ -93,9 +93,11 @@
     "number_helpers": "^0.1.1",
     "patternfly": "^3.58.0",
-    "patternfly-react": "^2.30.6",
+    "patternfly-react": "^2.32.4",
```
https://github.com/theforeman/foreman-packaging/pull/3752

ares replied:
changes requested there, waiting on packaging ack here, I retested and functionality wise it's ready :+1:

ares reviewed (dismissed):
Works nicely now, waiting just on packaging ack

mmoll reviewed (dismissed):
Approving, @glekner please look out for comments in the RPM PRs. Leaving the merge up to @ohadlevy.

mmoll commented:
@glekner one last rebase please :)

glekner commented:
rebased

ohadlevy commented:
[test foreman]

ohadlevy commented on package.json:
```diff
@@ -96,6 +96,7 @@
     "patternfly-react": "^2.34.1",
```
seems like you are missing patternfly-react-extensions after the rebase?

mmoll reviewed (approved):
RPM PRs are open

mmoll commented:
merged, thanks @glekner!

tbrisker commented:
RPM building is now failing, the packaging prs weren't merged and i'm now confused which one of them should be merged.
